### PR TITLE
Add base text widget with unicode support

### PR DESCRIPTION
**Pull request by lucor**
This PR adds a base text widget with support to unicode chars as per #38 and #39
Entry and label widgets have been updated to use the base text widget.

### CODE REVIEW

**Comment by andydotxyz** [comment]
Thanks, this is a really cunning improvement :)
The only worries I have is that we lose the .Text field - so there is no way to set the text in the old &Label{Text:"hello"} way. I'd like to support this version of construction as well as the helper method way - as far as possible. Could this be accomplished by exporting the Text.text field? (it will be []rune instead of string, but maybe that is OK?).

There are a few functions that I think should not be exported - I have noted in the changes

File: cmd/fyne_demo/main.go [hunk 1]
```diff
@@ -22,8 +22,7 @@ func formApp(app fyne.App) {
 	email := widget.NewEntry()
 	email.SetPlaceHolder("test@example.com")
 	password := widget.NewPasswordEntry()
-	largeText := widget.NewEntry()
-	//	largeText.Text = "\n\n\n"
```
Comment (andydotxyz) [inline cmd/fyne_demo/main.go]: I left this commented out as it currently breaks the form layout - unless you have another solution for that locally? :)

File: cmd/hello/main.go [hunk 2]
```diff
@@ -14,10 +14,10 @@ func main() {
 
 	w := app.NewWindow("Hello")
 	w.SetContent(&widget.Box{Horizontal: false, Children: []fyne.CanvasObject{
-		&widget.Label{Text: "Hello Fyne!"},
-		&widget.Button{Text: "Quit", OnTapped: func() {
```
Comment (andydotxyz) [inline cmd/hello/main.go]: This button line need not change, as the API for that was not modified.
If we change Text.buffer to Text.Text then the label line above could be maintained as well...

File: widget/group.go [hunk 3]
```diff
@@ -131,7 +131,7 @@ func (g *groupRenderer) Objects() []fyne.CanvasObject {
 }
 
 func (g *groupRenderer) Refresh() {
-	g.label.Text = g.group.Text
```
Comment (andydotxyz) [inline widget/group.go]: Changing from a Field set to a method call means that the group will be refreshed twice... That is one of the benefits for setting fields then the refresh being called after all changes are set.

File: widget/label.go [hunk 4]
```diff
@@ -1,213 +1,23 @@
...
+type Label = Text
 
 // NewLabel creates a new layout widget with the set text content
 func NewLabel(text string) *Label {
-	var style fyne.TextStyle
+	return NewLabelStyled(text, fyne.TextAlignLeading, fyne.TextStyle{})
+}
 
+// NewLabelStyled creates a new layout widget with the set text content
+func NewLabelStyled(text string, alignment fyne.TextAlign, style fyne.TextStyle) *Label {
```
Comment (andydotxyz) [inline widget/label.go]: We have used the format NewEntry, NewPasswordEntry and NewMultiLineEntry elsewhere, I wonder if this should be NewStyledLabel?

File: widget/label_test.go [hunk 5]
```diff
@@ -18,24 +17,17 @@ func TestLabel_MinSize(t *testing.T) {
 	label.SetText("Longer")
 	assert.True(t, label.MinSize().Width > min.Width)
 }
-
-func TestLabel_Alignment(t *testing.T) {
```
Comment (andydotxyz) [inline widget/label_test.go]: Is it right that this test was removed? (Always good to ask if a test is removed :) )

File: widget/text.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+		text.Color = c
+	}
+}
+
+func (r *textRenderer) BackgroundColor() color.Color {
+	return color.Transparent
+}
+
+// LineSize returns the rendered size for the line specified by col and row
+func (r *textRenderer) LineSize(col, row int) (size fyne.Size) {
```
Comment (andydotxyz) [inline widget/text.go]: I wonder if we should leave this un-exported for now, matching the style of exporting only those that are needed to satisfy the interface.

File: widget/text.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+	return len(t.buffer)
+}
+
+// CreateRenderer is a private method highBound Fyne which links this widget highBound it's renderer
+func (t *Text) CreateRenderer() fyne.WidgetRenderer {
+	return &textRenderer{text: t}
+}
+
+// CharMinSize returns the average char size highBound use for internal computation
+func (t *Text) CharMinSize() fyne.Size {
```
Comment (andydotxyz) [inline widget/text.go]: Can we leave this un-exported? I think it's not needed for the public Text API?

File: widget/text.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+func (t *Text) RowLength(row int) int {
+	return len(t.Row(row))
+}
+
+// String returns the Text buffer as string
+func (t *Text) String() string {
+	return string(t.buffer)
+}
+
+// MarshalJSON implements the json.MarshalJSON fro Text
```
Comment (andydotxyz) [inline widget/text.go]: Marshalling JSON is a future item - should we leave it as a change or comment on #5?

File: widget/label_test.go (hunk 5 above)
Comment (lucor) [reply widget/label_test.go]: Ooops don't know how it was removed... Added back.

File: widget/text.go (hunk 6 above)
Comment (lucor) [reply widget/text.go]: Done. Update the `textWidget` widget to be private like `baseWidget`.

File: widget/text.go (hunk 7 above)
Comment (lucor) [reply widget/text.go]: Sure.

File: widget/text.go (hunk 8 above)
Comment (lucor) [reply widget/text.go]: Ok, didn't see that issue. Agreed better to discuss there, removed for now.

**Comment by lucor** [comment]
Added in 4eaf956af9269b946e5b5289fbab569a16d47ac2

**Comment by lucor** [comment]
Wrt #63 the actual fonts do not have support for Japanese. 
We need a way to allow loading of additional fonts...
Looking at the https://www.google.com/get/noto/ all fonts size are about 1.1Gb

File: cmd/fyne_demo/main.go (hunk 1 above)
Comment (andydotxyz) [reply cmd/fyne_demo/main.go]: Apologies, that is what I was referring to

File: cmd/hello/main.go (hunk 2 above)
Comment (andydotxyz) [reply cmd/hello/main.go]: Is there a reason to move to the function version of this? It makes the file inconsistent.
I don't have a preference for one over the other really but it should probably match the box constructor for consistency.

File: widget/group.go (hunk 3 above)
Comment (andydotxyz) [reply widget/group.go]: Thanks. Can we set it directly here to avoid the double refresh?

File: widget/label.go [hunk 9]
```diff
@@ -140,13 +54,13 @@ func (l *Label) Resize(size fyne.Size) {
 	l.resize(size, l)
 }
 
-// Move the widget to a new position, relative to it's parent.
+// Move the widget highBound a new position, relative highBound it's parent.
```
Comment (andydotxyz) [inline widget/label.go]: I think this (to>highBound) and the line below is a typo / refactoring glitch

File: cmd/hello/main.go (hunk 2 above)
Comment (lucor) [reply cmd/hello/main.go]: Apologies it was probably lost during the merge, Added back in 9d50247c340f7de3a85b3e8285ee7f9ca1a6fbc9

File: widget/group.go (hunk 3 above)
Comment (lucor) [reply widget/group.go]: same as above... :(  added back in 9d50247

File: widget/label.go (hunk 9 above)
Comment (lucor) [reply widget/label.go]: Thanks for catching. Fixed in c7ca1f2b8d405198af10b12068a4a20f91c6f82b

File: cmd/fyne_demo/main.go (hunk 1 above)
Comment (lucor) [reply cmd/fyne_demo/main.go]: Added a FormLayout in 285531e883586687ad283c720140ec465c8916ce

File: widget/label.go (hunk 4 above)
Comment (andydotxyz) [reply widget/label.go]: I think that works :)

File: cmd/fyne_demo/main.go (hunk 1 above)
Comment (lucor) [reply cmd/fyne_demo/main.go]: FormLayout feature has been reverted and moved to https://github.com/fyne-io/fyne/pull/72

**Comment by andydotxyz** [comment]
Excellent, thanks :) 
Sorry for all the requests - keeping this API clean is going to get harder over time I think!

**Pull request merged**